server/src/index.ts

```diff
@@ -18,17 +18,61 @@ const io = new Server(server, {
   },
 });
 
+// Authentication codes to be returned to the client.
+// Needs to be in sync with the frontend enum.
+enum AuthenticationEnum {
+  Success = 0,
+  IncorrectPassword = 401,
+}
+
+// Holds info about all existing rooms.
+const rooms: { [key: string]: { numUsers: number; passHash: string } } = {};
+
+const getHash = (prehash: string) => {
+  return createHash("sha256").update(prehash).digest("hex");
+};
+
+// Function that returns the default hash (when a user provides no password).
+// Needs to be in agreement with the frontend.
+const getDefaultHash = (roomName: string) => {
+  return getHash(roomName + getHash(roomName));
+};
+
+const authenticate = (roomName: string, pass: string) => {
+  const prehash = roomName + pass;
+  const hash = getHash(prehash);
+  const room = rooms[roomName];
+
+  // If room doesn't exist, return.
+  if (!room) {
+    rooms[roomName] = { numUsers: 0, passHash: hash };
+    return AuthenticationEnum.Success;
+  } else {
+    const roomHash = room.passHash;
+    // If room password is set, authenticate, and if not, set password and allow connection.
+    if (roomHash) {
+      if (roomHash === hash) return AuthenticationEnum.Success;
+      else return AuthenticationEnum.IncorrectPassword;
+    }
+  }
+};
+
+const getRoomName = (nsp: string) => {
+  return encodeURIComponent(path.basename(nsp));
+};
+
 app.get("/", (req, res) => {
-  res.send({ body: "Hello world, 3" });
+  res.send({ body: "Hello world, 4" });
 });
 
-const rooms: { [key: string]: number } = {};
-
-const users: object[] = [];
 io.on("connection", (socket) => {
   socket.on("NUM_USERS", (nspName) => {
-    const roomName = stripTrailingSlash(nspName)
-    socket.emit("NUM_USERS", rooms[roomName]);
+    const roomName = getRoomName(nspName);
+    if (rooms[roomName]) {
+      socket.emit("NUM_USERS", rooms[roomName].numUsers);
```

Comment:
Will make those channel names into constant in #321

Reply:
Sounds good👍